### PR TITLE
Improve UnitCL pow testing.

### PR DESCRIPTION
**Pull request by hvdijk**
# Overview

Improve UnitCL pow testing.

# Reason for change

We single out the value 0x39f6 in our implementation of log2_extended_precision, so in tests that use that function, we should ensure that that value is always tested.

# Description of change

This adds an overridable `GetEdgeCases()` function, defaulting to the `InputGenerator::half_edge_cases` we previously used, but overridden for `HalfMathBuiltinsPow` to additionally add `0x39f6` as an edge case.

Tested by locally reverting commit 90cf2faccb27adb19ce330bbe7c7ec62c53c7018 and verifying that UnitCL reliably picks up the error.

# Anything else we should know?

*If there's any other relevant information we should know that may help us in
understanding and verifying your patch, please include it here.*

# Checklist

* Read and follow the project [Code of Conduct](https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/CODE_OF_CONDUCT.md).
* Make sure the project builds successfully with your changes.
* Run relevant testing locally to avoid regressions.
* Run [clang-format-16](https://clang.llvm.org/docs/ClangFormat.html) (the most
  recent version available through `pip`) on all modified code.

### CODE REVIEW

File: source/cl/test/UnitCL/source/ktst_precision.cpp [hunk 1]
```diff
@@ -413,6 +413,17 @@ UCL_EXECUTION_TEST_SUITE_P(HalfOperatorTest, testing::Values(OPENCL_C),
 
 using HalfMathBuiltins = HalfParamExecution;
 
+struct HalfMathBuiltinsPow : HalfMathBuiltins {
+  const std::vector<cl_ushort> &GetEdgeCases() const override {
+    static std::vector<cl_ushort> EdgeCases = [&] {
```
Comment (frasercrmck) [inline source/cl/test/UnitCL/source/ktst_precision.cpp:418]: How come this is done in a lambda?

File: source/cl/test/UnitCL/include/kts/execution.h [hunk 2]
```diff
@@ -347,6 +347,9 @@ class HalfParamExecution : public ExecutionWithParam<unsigned> {
   /// @brief Returns true if parameter at index has a scalar type
   /// @param[in] Index of parameter to check
   bool IsArgScalar(unsigned index) const;
+
+  /// @brief Returns a vector of edge cases that need extra testing.
+  virtual const std::vector<cl_ushort> &GetEdgeCases() const;
```
Comment (coldav) [inline source/cl/test/UnitCL/include/kts/execution.h:352]: is it really worth doing as a reference? It could be prone to people introducing bugs. Would it not be simpler to just add the edge cases on request?

File: source/cl/test/UnitCL/source/ktst_precision.cpp (hunk 1 above)
Comment (coldav) [reply source/cl/test/UnitCL/source/ktst_precision.cpp:418]: presumably to do it all in one static variable line.

**Review by coldav — APPROVED** [review]

File: source/cl/test/UnitCL/source/ktst_precision.cpp (hunk 1 above)
Comment (hvdijk) [reply source/cl/test/UnitCL/source/ktst_precision.cpp:418]: Yes, this is to directly construct `EdgeCases` as the vector we want, rather than first constructing it and then adding an element, because this way we are guaranteed the initialization happens exactly once without needing extra guards.

File: source/cl/test/UnitCL/include/kts/execution.h (hunk 2 above)
Comment (hvdijk) [reply source/cl/test/UnitCL/include/kts/execution.h:352]: I don't quite get what you mean here, could you elaborate?

File: source/cl/test/UnitCL/source/ktst_precision.cpp (hunk 1 above)
Comment (frasercrmck) [reply source/cl/test/UnitCL/source/ktst_precision.cpp:418]: Fair enough, yeah.

File: source/cl/test/UnitCL/include/kts/execution.h (hunk 2 above)
Comment (coldav) [reply source/cl/test/UnitCL/include/kts/execution.h:352]: I mean it can be called as is, it was more the reference, I'm just not sure that creating the vector at the time would really be noticeable time-wise

Comment (hvdijk) [reply source/cl/test/UnitCL/include/kts/execution.h:352]: True, this isn't performance-critical. I did check though that the risk you mentioned, of accidentally returning a local variable, is detected by the compiler. It is, it results in "error: reference to local variable ‘EdgeCases’ returned". It would be a warning, but we build with `-Werror` anyway. Because of that, I think the risk of accidental wrong use is low enough that we don't need to worry about it.

Comment (coldav) [reply source/cl/test/UnitCL/include/kts/execution.h:352]: I agree it does, feel free to merge as is.
